File: 4I504 - Compilation/src/universal_machine.rb
require_relative "reading"

module OPCODES
  CMV = 0
  AIND = 1
  AAMD = 2
  ADD = 3
  MUL = 4
  DIV = 5
  NAND = 6
  HALT = 7
  ALLOC = 8
  ABANDON = 9
  OUT = 10
  IN = 11
  LPROG = 12
  ORTH = 13
end

class UniversalMachine
  def get_opc(num)
    return (num >> 28) & 0b1111
  end

  def next_instr
    @val = @memory[0][@pc]
    @a = (@val >> 6) & 0b111
    @b = (@val >> 3) & 0b111
    @c = (@val >> 0) & 0b111
    @opc = get_opc(@memory[0][@pc])
    @pc = @pc + 1
  end

  def exec_vm
    @registers = [0, 0, 0, 0, 0, 0, 0, 0]
    @memory = []
    @memory << (read_prog ARGV[0])

    @pc = 0
    @val = 0
    @a = 0
    @b = 0
    @c = 0
    @opc = -1

    loop do
      next_instr
      case @opc
      when OPCODES::CMV
        @registers[@a] = @registers[@b] if @registers[@c] != 0
      when OPCODES::AIND
        @registers[@a] = @memory[@registers[@b]][@registers[@c]]
      when OPCODES::AAMD
        @memory[@registers[@a]][@registers[@b]] = @registers[@c]
      when OPCODES::ADD
        @registers[@a] = (@registers[@b] + @registers[@c]) % 4294967296
      when OPCODES::MUL
        @registers[@a] = (@registers[@b] * @registers[@c]) % 4294967296
      when OPCODES::DIV
        @registers[@a] = @registers[@b] / @registers[@c] if @registers[@c] != 0
      when OPCODES::NAND
        @registers[@a] = ~(@registers[@b] & @registers[@c])
        @registers[@a] += 2**32 if @registers[@a] < 0
      when OPCODES::HALT
        return
      when OPCODES::ALLOC
        temp = -1
        for i in 0...@memory.size
          if @memory[i] == nil
            temp = i
            break
          end
        end

        if temp == -1
          @memory << Array.new(@registers[@c], 0)
          @registers[@b] = @memory.size - 1
        else
          @memory[temp] = Array.new(@registers[@c], 0)
          @registers[@b] = temp
        end
      when OPCODES::ABANDON
        @memory[@registers[@c]] = nil
      when OPCODES::OUT
        if @registers[@c] > 255
          puts "Error out..."
          return
        end
        print @registers[@c].chr
      when OPCODES::IN
        input = getc
        if getc == nil
          @registers[@c] = 0xFFFFFFFF
        else
          @registers[@c] = input.ord
        end
      when OPCODES::LPROG
        if @registers[@b] != 0
          temp = @memory[@registers[@b]].dup
          @memory[0] = temp
        end
        @pc = @registers[@c]
      when OPCODES::ORTH
        @registers[(@val >> 25) & 0b111] = @val & 0x1FFFFFF
        @registers[(@val >> 25) & 0b111] += 2**32 if @registers[(@val >> 25) & 0b111] < 0
      end
    end
  end
end

vm = UniversalMachine.new
vm.exec_vm

File: 4I504 - Compilation/src/universal_machine.c
#include "universal_machine.h"

// Opcode.
#define get_opc(num) (num >> 28) & 0b1111

// Computed GOTO.
#define next()						\
  do {							\
    val = arrays[0].code[pc];				\
    a = (val >> 6) & 0b111;				\
    b = (val >> 3) & 0b111;				\
    c = (val >> 0) & 0b111;				\
    goto *opcodes[get_opc(arrays[0].code[pc++])];	\
  } while(0)
#define stop() goto end

// Registers.
unsigned int reg[] = { 0, 0, 0, 0, 0, 0, 0, 0 };

// Get the opcode.

#define debug(opcode) ;//printf("%s, pc = %x,\n\ta = %d, b = %d, c = %d \n", opcode, pc, reg[a], reg[b], reg[c])

// LET'S GO !
void exec_vm(Code program) {
  unsigned int size = 10, arr = 1;
  Code *arrays = malloc(sizeof *arrays * size);
  for (unsigned int i = 0; i < size; i++)
    arrays[i].size = -1;
  arrays[0] = program;

  static void* opcodes[] = {
    &&cmv,     &&aind,    &&aamd, &&add,
    &&mul,     &&div,     &&nand, &&halt,
    &&alloc,   &&abandon, &&out,  &&in,
    &&lprog,   &&orth
  };

  unsigned int pc = 0;
  unsigned int val = 0;
  unsigned int a, b, c;

  next();

 cmv:
  debug("cmv");
  if (reg[c] != 0)
    reg[a] = reg[b];
  next();

 aind:
  debug("aind");
  printf("%d\n", reg[b]);
  reg[a] = arrays[reg[b]].code[reg[c]];
  next();

 aamd:
  debug("aamd");
  arrays[reg[a]].code[reg[b]] = reg[c];
  next();

 add:
  debug("add");
  reg[a] = (reg[b] + reg[c]) % 4294967296; // 2**32.
  next();

 mul:
  debug("mul");
  reg[a] = (reg[b] * reg[c]) % 4294967296;
  next();

 div:
  debug("div");
  if (reg[c] != 0)
    //printf("%ld %ld %08x\n", reg[a], reg[b], (unsigned int) reg[c]);
    //printf("%d\n", reg[b] / reg[c]);
    reg[a] = ((unsigned int) reg[b] / (unsigned int) reg[c]);
  next();

 nand:
  debug("nand");
  reg[a] = ~(reg[b] & reg[c]);
  next();

 halt:
  debug("halt");
  stop();

 alloc:
  debug("alloc");

  if (size < arr) {
    Code *arrays_p;
    size = size * 2;
    if ((arrays_p = realloc(arrays, size)) != NULL) {
      free(arrays);
      arrays = arrays_p;
    } else {
      fprintf(stderr, "Realloc bug.\n");
      exit(EXIT_FAILURE);
    }
  }

  Code new;
  new.code = malloc(sizeof *(new.code) * reg[c]);
  new.size = reg[c];
  for (unsigned int i = 0; i < reg[c]; i++)
    new.code[i] = 0;
  arrays[size] = new;
  reg[b] = size++;

  next();

 abandon:
  debug("abandon");
  free(arrays[reg[c]].code);
  arrays[reg[c]].size = -1;
  next();

 out:
  debug("out");
  if (reg[c] > 255) {
    fprintf(stderr, "Output : bad number.\n");
    exit(EXIT_FAILURE);
  }
  printf("%c", (unsigned char) reg[c]);
  next();

 in:
  printf("in");
  {
    char input;
    if ((input = getc(stdin)) == EOF)
      reg[c] = 0xFFFFFFFF;
    else
      reg[c] = input;
    next();
  }
 lprog:
  debug("lprog");
  if (reg[b] != 0) {
    Code temp;
    temp.size = arrays[reg[b]].size;
    temp.code = malloc(sizeof *(temp.code) * temp.size);
    //printf("Ici\n");
    for (int i = 0; i < temp.size; i++)
      temp.code[i] = arrays[reg[b]].code[i];
    free(arrays[0].code);
    //printf("Ou la.\n");
    arrays[0] = temp;
  }
  pc = reg[c];

  next();

 orth:
  debug("orth");
  reg[(val >> 25) & 0b111] = val & 0x1FFFFFF;
  next();

 end:
  return;
}
